src/stdlib/slice.rs

```diff
@@ -109,6 +110,58 @@ struct SliceFn {
     end: Option<Box<dyn Expression>>,
 }
 
+#[allow(unused)]
+/// This function is an example of what we can do for future functions.
+/// For backwards compatibility, we cannot use this to improve existing behavior.
+/// After going through the deprecation process, we can use this to produce
+/// a more accurate type definition and make more `slice` uses infallible.
+fn determine_fallibility(
+    value: &dyn Expression,
+    start: &dyn Expression,
+    end: Option<&dyn Expression>,
+    state: &state::TypeState,
+) -> Fallibility {
+    let array = match value.resolve_constant(state) {
+        None => return Fallibility::AlwaysFails,
```

Comment:
This is wrong. If `resolve_constant` returns `None`, that just means the value is unknown at compile time, but it's certainly possible for it to be successful still. It seems like there are similar issues elsewhere too.

Reply:
In general I feel like we can add several helper functions to simplify this logic. The `determine_fallibility` function seems quite verbose right now. I'll try suggesting some if you can't come up with any (after the issue above is fixed)

Reply:
Thanks for the suggestions. It all makes sense. I will do this another PR in order to unblock this one.